lib/grandpa/finalization.go

```diff
@@ -0,0 +1,493 @@
+// Copyright 2022 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package grandpa
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/ChainSafe/gossamer/dot/telemetry"
+)
+
+type ephemeralService interface {
+	Start() error
+	Stop() error
+}
+
+type finalizationHandler struct {
+	servicesLock       sync.Mutex
+	finalizationEngine ephemeralService
+	votingRound        ephemeralService
+
+	newServices    func() (engine, voting ephemeralService)
+	timeoutStop    time.Duration
+	initiateRound  func() error
+	observableErrs chan error
+	stopCh         chan struct{}
+	handlerDone    chan struct{}
+}
+
+func newFinalizationHandler(service *Service) *finalizationHandler {
+	// builder is a constructor function which takes care to instantiate
+	// and return the services needed to finalize a round, those services
+	// are ephemeral services with a lifetime of a round
+	builder := func() (engine, voting ephemeralService) {
+		finalizationEngine := newFinalizationEngine(service)
+		votingRound := newHandleVotingRound(service, finalizationEngine.actionCh)
+		return finalizationEngine, votingRound
+	}
+
+	return &finalizationHandler{
+		newServices: builder,
+
+		timeoutStop:    5 * time.Second,
+		initiateRound:  service.initiateRound,
+		observableErrs: make(chan error),
+		stopCh:         make(chan struct{}),
+		handlerDone:    make(chan struct{}),
+	}
+}
+
+func (fh *finalizationHandler) Start() (errsCh <-chan error, err error) {
+	go fh.start()
+	return fh.observableErrs, nil
+}
+
+func (fh *finalizationHandler) start() {
+	defer close(fh.handlerDone)
+
+	for {
+		select {
+		case <-fh.stopCh:
+			return
+		default:
+		}
+
+		err := fh.initiateRound()
+		if err != nil {
+			fh.observableErrs <- fmt.Errorf("initiating round: %w", err)
+			return
+		}
+
+		err = fh.waitServices()
+		if err != nil {
+			fh.observableErrs <- err
+			return
+		}
+	}
+}
+
+func (fh *finalizationHandler) stop() (err error) {
+	fh.servicesLock.Lock()
+	defer fh.servicesLock.Unlock()
+
+	stopWg := new(sync.WaitGroup)
+	stopWg.Add(2)
+
+	errs := [2]error{}
+	go func() {
+		defer stopWg.Done()
+		errs[0] = fh.finalizationEngine.Stop()
+	}()
+
+	go func() {
+		defer stopWg.Done()
+		errs[1] = fh.votingRound.Stop()
+	}()
+
+	stopWg.Wait()
+
+	if errs[0] != nil {
+		return errs[0]
+	}
+
+	if errs[1] != nil {
+		return errs[1]
+	}
```

Comment:
Also we should *ideally* wrap one of the two since we cannot wrap 2 errors :cry: 

Something like `fmt.Errorf("both failed: %s; %w", errA, errB)`... or define a new sentinel error `ErrBothFailed` that you wrap with `fmt.Errorf("%w: %s; %s", ErrBothFailed, errA, errB)`... probably overthinking it a bit though :smile: